primitives/aleph-header-chain/src/aleph_justification.rs

```diff
@@ -0,0 +1,266 @@
+use codec::{Decode, Encode, Input};
+use frame_support::PalletError;
+use sp_runtime::{traits::Header as HeaderT, RuntimeAppPublic, RuntimeDebug};
+use sp_std::vec::Vec;
+
+use crate::{AuthoritySet, AuthoritySignature};
+use scale_info::TypeInfo;
+
+#[derive(TypeInfo, PartialEq, Eq, Clone, Debug, Decode, Encode)]
+pub struct Signature(AuthoritySignature);
+
+pub type SignatureSet = Vec<Option<Signature>>;
+
+/// A proof of block finality, currently in the form of a sufficiently long list of signatures or a
+/// sudo signature of a block for emergency finalization.
+#[derive(TypeInfo, Clone, Encode, Decode, Debug, PartialEq, Eq)]
+pub enum AlephJustification {
+	CommitteeMultisignature(SignatureSet),
+	EmergencySignature(AuthoritySignature),
+}
+
+#[derive(Eq, Encode, Decode, PartialEq, Debug, Copy, Clone)]
+pub struct Version(pub u16);
+
+/// Actual on-chain justification format.
+#[derive(Clone, Encode, Debug, PartialEq, Eq)]
+pub struct VersionedAlephJustification {
+	version: Version,
+	num_bytes: u16,
+	justification: AlephJustification,
+}
+
+impl Decode for VersionedAlephJustification {
+	fn decode<I: Input>(input: &mut I) -> Result<Self, codec::Error> {
+		let version = Version::decode(input)?;
+		let num_bytes = u16::decode(input)?;
+		let justification = match version {
+			Version(3) => AlephJustification::decode(input)?,
+			_ => return Err(codec::Error::from("Unsupported justification version")),
+		};
+		Ok(VersionedAlephJustification { version, num_bytes, justification })
+	}
+}
+
+impl From<VersionedAlephJustification> for AlephJustification {
+	fn from(justification: VersionedAlephJustification) -> Self {
+		justification.justification
+	}
+}
+
+#[derive(Eq, RuntimeDebug, PartialEq, Encode, Decode, TypeInfo, PalletError)]
+pub enum Error {
+	JustificationNotDecodable,
+	NotEnoughCorrectSignatures,
+	InvalidIndex,
+	EmergencyFinalizerUsed,
+}
+
+pub fn verify_justification<Header: HeaderT>(
+	authority_set: &AuthoritySet,
+	header: &Header,
+	justification: &AlephJustification,
+) -> Result<(), Error> {
+	match justification {
+		AlephJustification::CommitteeMultisignature(signature_set) => {
+			let mut signature_count = 0;
+			let enumerated_signatures = signature_set
+				.iter()
+				.enumerate()
+				.filter_map(|(idx, maybe_value)| Some((idx, maybe_value.as_ref()?)));
```

Comment:
Suggestion: Maybe something like:

```suggestion
				.filter_map(|(idx, maybe_value)| Some((idx, maybe_value.as_ref()?)))
				.filter(|(idx, signature)| is_correct(&idx, &signature, &authority_set))
				.count()
```
?